pkg/virtualKubelet/reflection/workload/podns.go

```diff
@@ -0,0 +1,446 @@
+
+	local, lerr := npr.localPods.Get(name)
+	utilruntime.Must(client.IgnoreNotFound(lerr))
+	localExists := !kerrors.IsNotFound(lerr)
+
+	remote, rerr := npr.remotePods.Get(name)
+	utilruntime.Must(client.IgnoreNotFound(rerr))
+	remoteExists := !kerrors.IsNotFound(rerr)
+
+	shadow, serr := npr.remoteShadowPods.Get(name)
+	utilruntime.Must(client.IgnoreNotFound(serr))
+	shadowExists := !kerrors.IsNotFound(serr)
+
+	tracer.Step("Retrieved the local and remote objects")
+
+	// Abort the reflection if the remote object is not managed by us, as we do not want to mutate others' objects.
+	if (remoteExists && !forge.IsReflected(remote)) || (shadowExists && !forge.IsReflected(shadow)) {
+		klog.Infof("Skipping reflection of local pod %q as remote already exists and is not managed by us", npr.LocalRef(name))
+		return nil
+	}
+	tracer.Step("Performed the sanity checks")
+
+	// The local pod does no longer exist. Ensure the shadowpod is absent from the remote cluster.
+	if !localExists {
+		defer tracer.Step("Ensured the absence of the remote object")
+		if !kerrors.IsNotFound(serr) {
+			klog.V(4).Infof("Deleting remote shadowpod %q, since local pod %q does no longer exist", npr.RemoteRef(name), npr.LocalRef(name))
+			return npr.DeleteRemote(ctx, npr.remoteShadowPodsClient, "ShadowPod", name, shadow.GetUID())
+		}
+
+		klog.V(4).Infof("Local pod %q and remote shadowpod %q both vanished", npr.LocalRef(name), npr.RemoteRef(name))
+		return nil
+	}
+
+	// The local pod is being terminated, hence mark also the remote one as to be deleted.
+	if !local.DeletionTimestamp.IsZero() {
+		// The remote pod has terminated, and we need to delete the local one.
+		if !remoteExists && !shadowExists {
+			defer tracer.Step("Ensured the absence of the local terminating object")
+
+			klog.V(4).Infof("Deleting terminating local pod %q, since remote %q does no longer exist", npr.LocalRef(name), npr.RemoteRef(name))
+			opts := metav1.NewDeleteOptions(0 /* trigger the effective deletion */)
+			opts.Preconditions = metav1.NewUIDPreconditions(string(local.GetUID()))
+			if err := npr.localPodsClient.Delete(ctx, name, *opts); err != nil && !kerrors.IsNotFound(err) {
+				klog.Errorf("Failed to delete local terminated pod %q: %v", npr.LocalRef(name), err)
+				return err
+			}
+
+			npr.ForgetPodInfo(name)
+			klog.Infof("Local pod %q successfully deleted", npr.LocalRef(name))
+			return nil
+		}
+
+		// The remote object is not yet terminating, trigger its deletion.
+		if shadowExists && shadow.DeletionTimestamp.IsZero() {
+			defer tracer.Step("Ensured the absence of the remote object")
+			klog.V(4).Infof("Deleting remote shadowpod %q, since local pod %q is terminating", npr.RemoteRef(name), npr.LocalRef(name))
+			return npr.DeleteRemote(ctx, npr.remoteShadowPodsClient, "ShadowPod", name, shadow.GetUID())
+		}
+
+		// The remote pod has already terminated.
+		if !remoteExists {
```

Comment:
Can we move this check in the HandleStatus method? 

It sounds like a precondition for the method to run. Indeed, the check is repeated https://github.com/liqotech/liqo/blob/e71366c6e7467a1a3ac13544d7259cec5cad6be8/pkg/virtualKubelet/reflection/workload/podns.go#L226-L233

Also, I don't think this path is actually tested. Checking for the remote pod existence inside the handlestatus method would also make testing it easier.